letsencrypt/proof_of_possession.py

```diff
@@ -61,6 +62,13 @@ def perform(self, achall):
             if cert_key == achall.hints.jwk:
                 return self._gen_response(achall, key)
 
+        if os.path.isdir(str(self.config.key_dir)):
+            for key_file in os.listdir(self.config.key_dir):
+                full_path = os.path.join(self.config.key_dir, key_file)
+                response = self._gen_response(achall, full_path)
+                if response is not False:
```

Comment:
So this doesn't quite work. The idea behind proof of possession is that the server gives the client a public key and the client has to sign a message using the corresponding private key.

`self._gen_response` simply takes a key and signs the message with no check that the correct private key was used. The result is that the code currently grabs the first private key it finds in `key_dir` and uses it to sign the message. What you need to do is load each private key as a JSON web key, get the corresponding public key, and compare it with the key given by the server. Only when you find a match can you generate a response and return it.

To do this nicely will require a bit of refactoring. Basically what you want to do is factor out the code that loads the key in `_gen_response` into a new function. You can then use this function both in `_gen_response` and in the code above to return a JWK object. You can then call `.public_key()` on that object to return the public key. After that, you need to compare it `achall.hints.jwk`. If there is a match, use the already loaded in key to generate the response.

Hopefully that all makes sense. Please follow up if you have any questions.